File: app/src/main/java/com/alex/morgan/bearlist/list/BearListPresenter.java
package com.alex.morgan.bearlist.list;

import com.alex.morgan.bearlist.Bear;

import java.util.Collection;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
class BearListPresenter implements BearListContract.Presenter, BearFetcher.Callback {

    private final BearFetcher bearFetcher;

    private BearListContract.View view;

    @Inject
    BearListPresenter(BearFetcher bearFetcher) {
        this.bearFetcher = bearFetcher;
    }

    @Override
    public void attachView(BearListContract.View view) {
        this.view = view;
        this.startLoading();
        bearFetcher.fetchBears(this, false);
    }

    @Override
    public void detachView() {
        this.view = null;
    }

    @Override
    public void onRefreshBears() {
        bearFetcher.fetchBears(this, true);
    }

    @Override
    public void onShowBearDetail(Bear bear) {
        //TODO
    }

    private void startLoading() {
        int loadingCounter = (int) System.currentTimeMillis() % 3;
        String loadingText;
        switch (loadingCounter) {
            case 0:
                loadingText = "Bear with me...";
                break;
            case 1:
                loadingText = "This should bearly take a second...";
                break;
            case 2:
                loadingText = "Fetching the bear necessities...";
                break;
            default:
                throw new RuntimeException();
        }

        view.showLoading(loadingText);
    }

    @Override
    public void onBears(Collection<Bear> bears) {
        view.showBears(bears);
    }

    @Override
    public void onError() {
        view.showError();
    }
}

File: app/src/main/java/com/alex/morgan/bearlist/list/BearListPresenter.kt
package com.alex.morgan.bearlist.list

import com.alex.morgan.bearlist.Bear
import com.alex.morgan.bearlist.app.ActivityScope
import dagger.Lazy
import javax.inject.Inject

@ActivityScope
internal class BearListPresenter @Inject
constructor(private val bearFetcher: BearFetcher, private val machine: Lazy<RubeGoldbergMachine>) :
    BearListContract.Presenter, BearFetcher.Callback {

    private var view: BearListContract.View? = null

    override fun attachView(view: BearListContract.View) {
        this.view = view
        this.startLoading()
        bearFetcher.fetchBears(this, false)
    }

    override fun detachView() {
        this.view = null
    }

    override fun onRefreshBears() {
        bearFetcher.fetchBears(this, true)
    }

    override fun onShowBearDetail(bear: Bear) {
        //TODO
    }

    override fun onActivateRubeGoldbergMachine() {
        machine.get().activate()
        view!!.showMessage("Done.")
    }

    private fun startLoading() {
        val loadingCounter = (System.currentTimeMillis() % 3).toInt()
        val loadingText: String
        when (loadingCounter) {
            0 -> loadingText = "Bear with me..."
            1 -> loadingText = "This should bearly take a second..."
            2 -> loadingText = "Fetching the bear necessities..."
            else -> throw RuntimeException("Loading counter was $loadingCounter")
        }

        view!!.showLoading(loadingText)
    }

    override fun onBears(bears: Collection<Bear>) {
        view!!.showBears(bears)
    }

    override fun onError() {
        view!!.showError()
    }
}
